File: packages/compiler/src/transforms/vIf.ts
import { DynamicFlag, IRNodeTypes } from '../ir'
import {
  createStructuralDirectiveTransform,
  type NodeTransform,
  type TransformContext,
} from '../transform'
import {
  createBranch,
  createCompilerError,
  createSimpleExpression,
  ErrorCodes,
  isConstantNode,
  isEmptyText,
  resolveDirective,
} from '../utils'
import type { JSXAttribute, JSXElement } from 'oxc-parser'

export const transformVIf: NodeTransform = createStructuralDirectiveTransform(
  ['if', 'else', 'else-if'],
  processIf,
)

export const transformedIfNode = new WeakMap()

export function processIf(
  node: JSXElement,
  attribute: JSXAttribute,
  context: TransformContext,
): (() => void) | undefined {
  const dir = resolveDirective(attribute, context)
  if (dir.name !== 'else' && (!dir.exp || !dir.exp.content.trim())) {
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_IF_NO_EXPRESSION, dir.loc),
    )
    dir.exp = createSimpleExpression(
      `true`,
      false,
      dir.exp ? dir.exp.ast : node,
    )
  }

  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
  transformedIfNode.set(node, dir)
  if (dir.name === 'if') {
    const id = context.reference()
    context.dynamic.flags |= DynamicFlag.INSERT
    const [branch, onExit] = createBranch(node, context)

    return () => {
      onExit()
      context.dynamic.operation = {
        type: IRNodeTypes.IF,
        id,
        condition: dir.exp!,
        positive: branch,
        once: context.inVOnce || isConstantNode(attribute.value!),
      }
    }
  } else {
    // check the adjacent v-if
    const siblingIf = getSiblingIf(context as TransformContext<JSXElement>)

    const siblings = context.parent && context.parent.dynamic.children
    let lastIfNode
    if (siblings) {
      let i = siblings.length
      while (i--) {
        if (
          siblings[i].operation &&
          siblings[i].operation!.type === IRNodeTypes.IF
        ) {
          lastIfNode = siblings[i].operation
          break
        }
      }
    }

    if (
      // check if v-if is the sibling node
      !siblingIf ||
      // check if IfNode is the last operation and get the root IfNode
      !lastIfNode ||
      lastIfNode.type !== IRNodeTypes.IF
    ) {
      context.options.onError(
        createCompilerError(ErrorCodes.X_V_ELSE_NO_ADJACENT_IF, node.range),
      )
      return
    }

    while (lastIfNode.negative && lastIfNode.negative.type === IRNodeTypes.IF) {
      lastIfNode = lastIfNode.negative
    }

    // Check if v-else was followed by v-else-if
    if (dir.name === 'else-if' && lastIfNode.negative) {
      context.options.onError(
        createCompilerError(ErrorCodes.X_V_ELSE_NO_ADJACENT_IF, node.range),
      )
    }

    const [branch, onExit] = createBranch(node, context)

    if (dir.name === 'else') {
      lastIfNode.negative = branch
    } else {
      lastIfNode.negative = {
        type: IRNodeTypes.IF,
        id: -1,
        condition: dir.exp!,
        positive: branch,
        once: context.inVOnce,
      }
    }

    return () => onExit()
  }
}

export function getSiblingIf(context: TransformContext<JSXElement>) {
  const parent = context.parent
  if (!parent) return

  const siblings = parent.node.children
  let sibling
  let i = siblings.indexOf(context.node)
  while (--i >= 0) {
    if (!isEmptyText(siblings[i])) {
      sibling = siblings[i]
      break
    }
  }

  if (
    sibling &&
    sibling.type === 'JSXElement' &&
    transformedIfNode.has(sibling)
  ) {
    return sibling
  }
}

File: packages/compiler-rs/src/transform/v_if.rs
use napi::{Either, bindgen_prelude::Either16};
use oxc_allocator::TakeIn;
use oxc_ast::ast::{Expression, JSXChild, JSXElement};
use oxc_span::SPAN;

use crate::{
  ir::index::{BlockIRNode, DynamicFlag, IRDynamicInfo, IfIRNode, SimpleExpressionNode},
  transform::{ContextNode, TransformContext},
  utils::{
    check::{is_constant_node, is_template},
    directive::resolve_directive,
    error::ErrorCodes,
    utils::{find_prop, find_prop_mut},
  },
};

pub fn transform_v_if<'a>(
  context_node: *mut ContextNode<'a>,
  context: &'a TransformContext<'a>,
  context_block: &'a mut BlockIRNode<'a>,
  _: &'a mut ContextNode<'a>,
) -> Option<Box<dyn FnOnce() + 'a>> {
  let Either::B(JSXChild::Element(node)) = (unsafe { &mut *context_node }) else {
    return None;
  };
  if is_template(node) && find_prop(node, Either::A("v-slot".to_string())).is_some() {
    return None;
  }
  let node = node as *mut oxc_allocator::Box<JSXElement>;

  let Some(dir) = find_prop_mut(
    unsafe { &mut *node },
    Either::B(vec![
      "v-if".to_string(),
      "v-else".to_string(),
      "v-else-if".to_string(),
    ]),
  ) else {
    return None;
  };
  let seen = &mut context.seen.borrow_mut();
  let start = dir.span.start;
  if seen.contains(&start) {
    return None;
  }
  seen.insert(start);

  let mut dir = resolve_directive(dir, context);
  if dir.name != "else"
    && (dir.exp.is_none() || dir.exp.as_ref().unwrap().content.trim().is_empty())
  {
    context.options.on_error.as_ref()(ErrorCodes::VIfNoExpression, dir.loc);
    dir.exp = Some(SimpleExpressionNode {
      content: "true".to_string(),
      is_static: false,
      loc: SPAN,
      ast: None,
    });
  }

  let dynamic = &mut context_block.dynamic;
  dynamic.flags |= DynamicFlag::NonTemplate as i32;

  if dir.name == "if" {
    let id = context.reference(dynamic);
    dynamic.flags |= DynamicFlag::Insert as i32;
    let block = context_block as *mut BlockIRNode;
    let exit_block = context.create_block(
      unsafe { &mut *context_node },
      unsafe { &mut *block },
      Expression::JSXElement(oxc_allocator::Box::new_in(
        unsafe { &mut *node }.take_in(context.allocator),
        context.allocator,
      )),
      None,
    );
    return Some(Box::new(move || {
      let block = exit_block();

      context_block.dynamic.operation = Some(Box::new(Either16::A(IfIRNode {
        id,
        positive: block,
        once: *context.in_v_once.borrow()
          || is_constant_node(&dir.exp.as_ref().unwrap().ast.as_deref()),
        condition: dir.exp.unwrap(),
        negative: None,
        anchor: None,
        parent: None,
      })));
    }));
  }

  let siblings = &mut context.parent_dynamic.borrow_mut().children;
  let mut last_if_node = None;
  if siblings.len() > 0 {
    let mut i = siblings.len();
    while i > 0 {
      i = i - 1;
      let sibling = siblings.get_mut(i).unwrap() as *mut IRDynamicInfo;
      if let Some(operation) = (unsafe { &mut *sibling }).operation.as_mut()
        && let Either16::A(operation) = operation.as_mut()
      {
        last_if_node = Some(operation);
        break;
      }
    }
  }

  // check if IfNode is the last operation and get the root IfNode
  let Some(mut last_if_node) = last_if_node else {
    context.options.on_error.as_ref()(ErrorCodes::VElseNoAdjacentIf, unsafe { &*node }.span);
    return None;
  };

  let mut last_if_node_ptr = last_if_node as *mut IfIRNode;
  while let Some(negative) = (unsafe { &mut *last_if_node_ptr }).negative.as_mut()
    && let Either::B(negative) = negative.as_mut()
  {
    last_if_node_ptr = negative as *mut IfIRNode;
  }
  last_if_node = unsafe { &mut *last_if_node_ptr };

  // Check if v-else was followed by v-else-if
  if dir.name == "else-if" && last_if_node.negative.is_some() {
    context.options.on_error.as_ref()(ErrorCodes::VElseNoAdjacentIf, dir.loc);
  };

  let exit_block = context.create_block(
    unsafe { &mut *context_node },
    context_block,
    Expression::JSXElement(oxc_allocator::Box::new_in(
      unsafe { &mut *node }.take_in(context.allocator),
      context.allocator,
    )),
    None,
  );

  Some(Box::new(move || {
    let block = exit_block();
    if dir.name == "else" {
      last_if_node.negative = Some(Box::new(Either::A(block)));
    } else {
      last_if_node.negative = Some(Box::new(Either::B(IfIRNode {
        id: -1,
        positive: block,
        once: *context.in_v_once.borrow()
          || is_constant_node(&dir.exp.as_ref().unwrap().ast.as_deref()),
        condition: dir.exp.unwrap(),
        parent: None,
        anchor: None,
        negative: None,
      })))
    }
  }))
}
